### PR TITLE
Dev/385 [POST] /api/groups/{groupId}/homework/{homeworkId}/progress && [POST]…

### LINKED ISSUE #385: [POST] /api/groups/{groupId}/homework/{homeworkId}/progress && [POST] /api/students/homework/{homeworkId}/progress
`[POST] /api/students/homework/{homeworkId}/progress`

RequestBody
[studentA@gmail.com, studentB@gmail.com]

ResponseBody
[reference](https://github.com/Judge-Girl/Judge-Girl/blob/4f05403d3766cf079c291ae134f9b16113cbbb8e/Spring-Boot/Spring-Boot-Academy/src/main/java/tw/waterball/judgegirl/springboot/academy/view/HomeworkProgress.java#L20)

--- 
`[POST] /api/groups/homework/{homeworkId}/progress`

RequestBody
[groupName-1, groupName-2]

---
image

### PR DESCRIPTION
**Pull request by sh910913**
resolve: #385

### CODE REVIEW

**Review by wallychen0705 — CHANGES_REQUESTED** [review]
Review done

**Review by wallychen0705 — CHANGES_REQUESTED** [review]
Review done

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review done! 😄 
To follow the rule of ubiquitous language, we don't have `question` in this context.
Please rename all `question` to `problem`.

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java [hunk 1]
```diff
@@ -276,5 +371,32 @@ private HomeworkProgress getHomeworkProgress(int studentId, int homeworkId) thro
...
+    private StudentView signUpStudent(String name, String email, String password) {
+        Student newStudent = new Student(name, email, password);
+        studentServiceDriver.addStudent(newStudent);
+        return toViewModel(newStudent);
+    }
+
+    private Group givenGroupWithMembers(String name, Integer... studentIds) {
+        Group group = groupRepository.save(new Group(name));
+        group.addMembers(mapToSet(studentIds, MemberId::new));
+        return groupRepository.save(group);
```
[the remark this answers is not in the document]
Comment (sh910913) [reply Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java:397]: It is not work.
got IllegalStateException: You can't save the group with its memberships at the same time.
image

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/GetGroupsHomeworkProgressUseCase.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    private final StudentServiceDriver studentServiceDriver;
+    private final GroupRepository groupRepository;
+    private final SubmissionServiceDriver submissionServiceDriver;
+
+    public GetGroupsHomeworkProgressUseCase(HomeworkRepository homeworkRepository, StudentServiceDriver studentServiceDriver, GroupRepository groupRepository, SubmissionServiceDriver submissionServiceDriver) {
+        super(homeworkRepository);
+        this.studentServiceDriver = studentServiceDriver;
+        this.groupRepository = groupRepository;
+        this.submissionServiceDriver = submissionServiceDriver;
+    }
```
[the remark this answers is not in the document]
Comment (Wally5077) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/GetGroupsHomeworkProgressUseCase.java]: @sh910913

[the remark this answers is not in the document]
Comment (sh910913) [reply Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/GetGroupsHomeworkProgressUseCase.java]: @Wally5077 這個已經有改過了喔。

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java [hunk 3]
```diff
@@ -159,12 +190,70 @@ public void testGetHomeworkProgress() throws Exception {
...
+
+    }
+
+    private void progressShouldHaveGrade(StudentsHomeworkProgress.StudentProgress progress, SubmissionView... bestRecords) {
+        for (SubmissionView bestRecord : bestRecords) {
+            var actualProblemScores = bestRecord.getVerdict() != null ? bestRecord.getVerdict().getTotalGrade() : 0;
+            var expectedProblemScores = progress.getProblemScores().get(bestRecord.getProblemId());
+            assertEquals(expectedProblemScores, actualProblemScores);
+        }
+    }
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java:248]: This should be amended.

**Review by wallychen0705 — CHANGES_REQUESTED** [review]
Review done

**Review by wallychen0705 — CHANGES_REQUESTED** [review]
Review done

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review done!

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/HomeworkProgressPresenter.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package tw.waterball.judgegirl.academy.domain.usecases.homework;
+
+import tw.waterball.judgegirl.primitives.Student;
+
+import java.util.List;
+
+/**
+ * @author sh910913@gmail.com
+ * @version 创建时间：2021/11/13 10:33 PM
```
Comment (Chaoyuuu) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/HomeworkProgressPresenter.java:9]: please remove this line. :rofl:

File: Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/StudentSubmissionRecord.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import tw.waterball.judgegirl.primitives.Student;
+import tw.waterball.judgegirl.submissionapi.views.SubmissionView;
+
+import java.util.List;
+
+/**
+ * @author sh910913@gmail.com
+ * @version 创建时间：2021/11/13 10:44 PM
```
Comment (Chaoyuuu) [inline Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/homework/StudentSubmissionRecord.java:13]: ditto

File: Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java [hunk 6]
```diff
@@ -159,20 +190,93 @@ public void testGetHomeworkProgress() throws Exception {
...
+
+        assertTrue(studentsHomeworkProgresses.containsKey(studentA.email));
+        assertTrue(studentsHomeworkProgresses.containsKey(studentB.email));
+        var progressA = studentsHomeworkProgresses.get(studentA.email);
+        var progressB = studentsHomeworkProgresses.get(studentB.email);
+        progressShouldHaveGrade(progressA, studentA, homework.problemIds);
+        progressShouldHaveGrade(progressB, studentB, homework.problemIds);
+    }
+
+    @DisplayName("Given a homework consists of two problems [p1, p2] and two students [A, B, C]," +
```
Comment (Chaoyuuu) [inline Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/HomeworkControllerTest.java:216]: 1. ... `and three students [A, B, C]`
2. you miss the context "group students to the group"

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review done. Thanks.

**Review by wallychen0705 — APPROVED** [review]
LGTM

**Review by Chaoyuuu — APPROVED** [review]
LGTM! Thanks! 😊

**Pull request merged**